lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,178 @@
+    nav_sections: dict[SectionHeader, list[StreamlitPage]]
+) -> list[StreamlitPage]:
+    page_list = []
+    for pgs in nav_sections.values():
+        for page in pgs:
+            page_list.append(page)
+
+    return page_list
+
+
+def nav_sections_to_page_dict(
+    nav_sections: dict[SectionHeader, list[StreamlitPage]]
+) -> dict[PageHash, PageInfo]:
+    d: dict[PageHash, PageInfo] = {}
+    for section_header in nav_sections:
+        for page in nav_sections[section_header]:
+            if isinstance(page._page, Path):
+                script_path = str(page._page)
+            else:
+                script_path = ""
+            d[page._script_hash] = {
+                "page_script_hash": page._script_hash,
+                "page_name": page.title,
+                "icon": page.icon,
+                "script_path": script_path,
+                "url_pathname": page.title.replace(" ", "_"),
+            }
+    return d
+
+
+def send_page_not_found(ctx: ScriptRunContext):
+    msg = ForwardMsg()
+    msg.page_not_found.page_name = ""
+    ctx.enqueue(msg)
+
+
+@gather_metrics("navigation")
+def navigation(
+    pages: list[StreamlitPage] | dict[str, list[StreamlitPage]],
+    *,
+    position: Literal["sidebar"] | Literal["hidden"] = "sidebar",
+) -> StreamlitPage | None:
+    """
+    Configure the available pages in a multipage app.
+
+    Call `st.navigation` in your main script with one or more pages defined by
+    `st.Page`. `st.navigation` returns the current page which can be executed
+    using `Page.run()`.
+
+    When using `st.navigation`, the main script is executed on every app rerun,
+    with the current page executed in-line using `Page.run()`. The set of
+    available pages can be updated with each rerun for dynamic navigation. In
+    this mode, the `pages/` folder is ignored.
+
+    By default, `st.navigation` draws the available pages in the side
+    navigation. This behavior can be changed using the `position=` keyword
+    argument.
+
+    Parameters
+    ----------
+    pages: list[st.Page] or dict[str, list[st.Page]]
+        A list of `st.Page` objects or a dictionary where the keys are section
+        headers and the values are lists of `st.Page` objects.
+
+    position: str
+        The position of the navigation menu. Can be "sidebar" or "hidden".
+
+    Example
+    -------
+    >>> import streamlit as st
+    >>> from pages import page1, page2
+    >>>
+    >>>	pg = st.navigation([st.Page(page1), st.Page(page2)])
+    >>>
+    >>>	st.title("My Awesome App")
+    >>> pg.run()
+    """
+    ctx = get_script_run_ctx()
+    if not ctx:
+        return None
+
+    nav_sections = {"": pages} if isinstance(pages, list) else pages
+    page_list = pages_from_nav_sections(nav_sections)
+
+    if len(page_list) == 0:
+        raise StreamlitAPIException(
+            "`st.navigation` must be called with at least one `st.Page`."
+        )
+
+    defaults = [page for page in page_list if page.default]
```

Comment:
I am wondering, whether this is something that can be returned already from `pages_from_nav_sections` in the line [above](https://github.com/streamlit/streamlit/pull/8673/files#diff-cde9e91039649c86b0e33690bf374e49cb91304e359cb681c2f4c4414e0d3859R118). Same with [these nested loops](https://github.com/streamlit/streamlit/pull/8673/files#diff-cde9e91039649c86b0e33690bf374e49cb91304e359cb681c2f4c4414e0d3859R142-R150) below which looks like the same loops as in `pages_from_nav_sections`. 
Not want to over-optimize, especially since the lists are likely very small - but maybe there is potential to merge the logic and have the loops only once, which might make the code smaller and thus easier to read.

Reply:
So I cleaned it up with the suggestion, but the protobuf population needs to be done after. This is due to the fact that we don't know the default page to set correctly in the protobuf until later.